pkg/controlplane/reconcilers/endpointsadapter.go

```diff
@@ -93,19 +93,8 @@ func (adapter *EndpointsAdapter) EnsureEndpointSliceFromEndpoints(namespace stri
 		return err
 	}
 
-	// required for transition from IP to IPv4 address type.
-	if currentEndpointSlice.AddressType != endpointSlice.AddressType {
-		err = adapter.endpointSliceClient.EndpointSlices(namespace).Delete(context.TODO(), endpointSlice.Name, metav1.DeleteOptions{})
-		if err != nil {
-			return err
-		}
-		_, err = adapter.endpointSliceClient.EndpointSlices(namespace).Create(context.TODO(), endpointSlice, metav1.CreateOptions{})
-		return err
-	}
-
 	if apiequality.Semantic.DeepEqual(currentEndpointSlice.Endpoints, endpointSlice.Endpoints) &&
-		apiequality.Semantic.DeepEqual(currentEndpointSlice.Ports, endpointSlice.Ports) &&
-		apiequality.Semantic.DeepEqual(currentEndpointSlice.Labels, endpointSlice.Labels) {
```

Comment:
we want to check the skip-mirror label is set to reconcile and avoid people remove it?
 related to https://github.com/kubernetes/kubernetes/issues/104252 ?

Reply:
Yeah I think all labels should be consistent here, is there any reason we'd want to remove this check?

Reply:
> we want to check the skip-mirror label is set

(The `skip-mirror` label is on the Endpoints, not the EndpointSlice; this check is about `kubernetes.io/service-name`.)

> to reconcile and avoid people remove it? related to #104252 ?

The existing apiserver code does not try to repair intentional sabotage. It only tries to fix objects that are in a previously-valid state.

> I think all labels should be consistent here, is there any reason we'd want to remove this check?

The apiserver always creates consistent labels already, so the only way they could be inconsistent would be if the administrator was manually trying to break it.

(We _could_ keep the Labels check here, but if so we should probably check Annotations, AddressType, and maybe a few other things too.)

Reply:
> (The skip-mirror label is on the Endpoints, not the EndpointSlice; this check is about kubernetes.io/service-name.)

duh :upside_down_face: , my bad

> > to reconcile and avoid people remove it? related to #104252 ?
> 
> The existing apiserver code does not try to repair intentional sabotage. It only tries to fix objects that are in a previously-valid state.

yeah, but it seems that the reason to not do that, was the multiport feature that you are going to remove ... should we solve this now?

Reply:
I don't know that he was talking about the _existing_ multi-port feature, but more about the problem of apiservers having different versions and/or configs in general. The apiservers have a protocol for reaching agreement on the set of endpoint IPs, but it is assumed/required that they already agree on what the value of every other Service/Endpoints field should be. If they don't, there will be edit wars if each apiserver regularly attempts to assert its own values.

Reply:
ok, let me write it down because I recognize I wasn't doing my homework correctly and I was confused and maybe we are talking about different reconcilers:

- The apiserver creates a service "kubernetes.default" and it must have endpoints so cluster members can reach it, it uses this setting to configure the IP address on the endpoint
```
	// PublicAddress is the IP address where members of the cluster (kubelet,
	// kube-proxy, services, etc.) can reach the GenericAPIServer.
	// If nil or 0.0.0.0, the host's default interface will be used.
	PublicAddress net.IP
```

- The apiserver has 3 types or reconcilers for these endpoints
```
	// MasterCountReconcilerType will select the original reconciler
	MasterCountReconcilerType Type = "master-count"
	// LeaseEndpointReconcilerType will select a storage based reconciler
	LeaseEndpointReconcilerType = "lease"
	// NoneEndpointReconcilerType will turn off the endpoint reconciler
	NoneEndpointReconcilerType = "none"
```

*None*: endpoints are generated externally, we don't care about this
*Master*: legacy, it depends on knowing beforehand the number of apiserver running
```
	// Number of masters running; all masters must be started with the
	// same value for this field. (Numbers > 1 currently untested.)
	MasterCount int
```
The reconciliation is based on the Endpoint object, it checks its own IP is in the object and the number of addresses is lower or equal to MasterCount


*Lease*: the default, each apiserver writes a lease directly in etcd and renews it periodically.
```
/registry/masterleases/172.18.0.4
/registry/masterleases/172.18.0.5
/registry/masterleases/172.18.0.6
```

The reconciliation is based on the IPs listed in the leases match the ones in the endpoint, and the lease TTL has renewed on time (there is an option to set the TTL expiration time)

I think we should start deprecating "Master" reconciler as it is easy to get conflicts, just if you don't set the MasterCount setting correctly.

If we consider only the Lease reconciler, I think that we can solve the problem of external actors modifying the apiserver endpoints.

Reply:
> I think we should start deprecating "Master" reconciler as it is easy to get conflicts, just if you don't set the MasterCount setting correctly.

Yeah, I already suggested in the dual stack apiserver KEP that we'd only support dual-stack apiservers with the `lease` reconciler.

> If we consider only the Lease reconciler, I think that we can solve the problem of external actors modifying the apiserver endpoints.

If we _only_ care about bad external actors we can solve it now, with either reconciler; each apiserver can just resync the ports value whenever it sees that the current value does not match what it thinks it should be.

The reason we don't do that (I assume) is because it fails when you're trying to change the config (or when a new version of kubernetes adds a new feature that changes the desired output; eg, if we want to add QUIC / HTTP3 support).

But yes, we can fix that case easily with the `lease` reconciler but not with the `master-count` reconciler; we just make the reconciler write out the ports to the lease objects as well as the IPs, and then adopt the rule that if all active masters agree on the value of the ports, then it should sync that value, but whenever there is disagreement between masters, it should leave the ports unchanged. (This matches the behavior we agreed would work best for dual-stack upgrade/rollback; the apiserver is dual-stack only when every apiserver writes out a dual-stack lease.)

(And we can't easily implement that same behavior with the `master-count` reconciler, because that reconciler doesn't have a separate set of objects where each apiserver can indicate its own theory of what the ports should be; the object that they use to communicate with each other is the `default/kubernetes` Endpoints object itself.)

But anyway, if we want to do this, this is a separate thing which is not incompatible with this PR. (And, as kind of implied by one of my other comments, in order to make dual-stack work, we'll have to flip the reconciler code to treat the EndpointSlices as the "real" output and generate the Endpoints from the EndpointSlices, rather than vice versa, so this is all going to get rewritten anyway.)

Reply:
> But anyway, if we want to do this, this is a separate thing which is not incompatible with this PR.

though also, I could just remove the "Remove some dead EndpointsAdapter code" commit from this PR, and have this be _just_ about making EndpointSlices required rather than optional

Reply:
that will be faster, that commit is the one creating doubts, for me, main concern is removing code that we can't easily predict the fallout ... this is a very sensitive change